diff --git a/src/utils/allocation.h b/src/utils/allocation.h
index 2cdd12242e..64b0f07350 100644
--- a/src/utils/allocation.h
+++ b/src/utils/allocation.h
@@ -35,6 +35,7 @@ class V8_EXPORT_PRIVATE Malloced {
 
 template <typename T>
 T* NewArray(size_t size) {
+  size = (size == 0) ? 1 : size;
   T* result = new (std::nothrow) T[size];
   if (result == nullptr) {
     V8::GetCurrentPlatform()->OnCriticalMemoryPressure();
