diff --git a/node.gyp b/node.gyp
index f18a0d58a8..5c4cc23ac0 100644
--- a/node.gyp
+++ b/node.gyp
@@ -818,6 +818,7 @@
             'Dbghelp',
             'Psapi',
             'Ws2_32',
+            'Winmm',
           ],
         }],
         [ 'node_use_etw=="true"', {
diff --git a/vcbuild.bat b/vcbuild.bat
index 4316658bb4..a455f60339 100644
--- a/vcbuild.bat
+++ b/vcbuild.bat
@@ -51,7 +51,7 @@ set noetw_msi_arg=
 set i18n_arg=
 set download_arg=
 set build_release=
-set configure_flags=
+set configure_flags=--no-browser-globals
 set build_addons=
 set dll=
 set enable_static=
